Automate secrets encryption and decryption

Edwin-Pau opened the pull request:
This python script allows the user to automate the encryption and decryption of the terraform.tfvars file so that it is ready to be used by Terraform for both types of deployments.

Commit:
Automate encryption using GCP KMS API

When using KMS encryption for the secrets inside the deployments'
tfvars files, users can now run a python script which will automate
the creation of the key ring, crypto key, and the encryption of the
tfvars secrets using the newly created crypto key.

Signed-off-by: Edwin Pau <epau@teradici.com>

Commit:
Automate the creation of tfvars with encrypted secrets

The python script that encrypts the secrets using the user provided
GCP service account will now create a a new terraform.tfvars file
with the encrypted secrets rendered into the appropriate fields.
The script overwrites the existing tfvars file with plaintext secrets,
and the result is a terraform.tfvars file that is ready to be used for
deployments (both single and multi-region are supported).

Signed-off-by: Edwin Pau <epau@teradici.com>

Commit:
Add support for dc-only and refactor script

Added the support for dc-only tfvars encryption. Refactored the code
for readability and rearranged functions so they are in alphabetical
order.

Signed-off-by: Edwin Pau <epau@teradici.com>

Commit:
Use argparse to allow users to set flags

Two flags are supported when executing this script. Encryption of
the tfvars is done by using the -e flag, while the decryption of
the tfvars is done by using the -d flag.

Signed-off-by: Edwin Pau <epau@teradici.com>

Commit:
Add decryption mode to GCP KMS wrapper script

Decryption mode is implemented to allow users a way to recover their
original terraform.tfvars which contains plaintext secrets. The
encrypted CAM credentials file is also decrypted back to its original
plaintext JSON credentials.

Signed-off-by: Edwin Pau <epau@teradici.com>

Commit:
Update README section on encrypting secrets

A section under "Encrypting Secrets" is added with more instructions
on using the encryption script, which automates the process of
encryption and decryption of secrets in terraform.tfvars, as well as
any text files specified under secrets as a path.

Signed-off-by: Edwin Pau <epau@teradici.com>

shermanyin commented on docs/README-gcp.md:
````diff
@@ -36,14 +36,22 @@ Login to Cloud Access Manager Admin Console at https://cam.teradici.com using a
+
+There are two modes for the script, use the ```-e``` flag for encryption of secrets or ```-d``` flag for decryption of secrets. Decryption mode allows the script to take your terraform.tfvars file containing encrypted secrets and decrypts them back to the original plaintext. The script rewrites the ciphertext version of terraform.tfvars file with the plaintext secrets again.
+
+The script also prompts the user for a deployment type, since there are three different deployment types. Enter ```0``` for single-connector, ```1``` for multi-region or ```1``` for dc-only terraform.tfvars.
````
0,1,1?

shermanyin commented on deployments/gcp/single-connector/terraform.tfvars.sample:
```diff
@@ -51,19 +65,29 @@ centos_std_instance_count = 0
-#   Secrets   #
-###############
+####################
+#      GCP KMS     #
```
I think leaving this section as "Secrets" is fine, since there is really only 2 actual config lines

shermanyin commented on deployments/gcp/single-connector/terraform.tfvars.sample:
```diff
@@ -51,19 +65,29 @@ centos_std_instance_count = 0
-# secrets. The GCP service account specified above in "gcp_service_account"
-# must be in the same project and have KMS decryptor permissions for this key.
+# For option 1, leave both keyring name and cryptokey name commented out.
+# For option 2, run the script located under tools using command "python3
```
shouldn't need "python3" in front

shermanyin commented on deployments/gcp/single-connector/terraform.tfvars.sample:
```diff
@@ -51,19 +65,29 @@ centos_std_instance_count = 0
+# your encrypted secrets and create an encrypted JSON for your CAM
+# credentials file. To set a custom keyring name and cryptokey name,
+# uncomment the two variables below and enter the new names, otherwise
+# a default name will be used by the script. The GCP service account 
```
don't need the last 2 sentences

shermanyin commented on deployments/gcp/multi-region/terraform.tfvars.sample:
```diff
@@ -35,6 +44,12 @@ cac_admin_ssh_pub_key_file = "~/.ssh/id_rsa.pub"
+####################
+#   Workstations   #
+####################
+# Specify the number of workstations created for each type.
```
remove. make sure the formatting are consistent between deployments

shermanyin commented on tools/gcp_kms_secrets.py:
```diff
@@ -0,0 +1,406 @@
+from google.cloud.kms_v1 import enums
+from google.oauth2       import service_account
+
+# Global variables
```
remove

Edwin-Pau replied:
I believe I need the service_account module to create credentials used for a client:
`credentials = service_account.Credentials.from_service_account_file(GCP_CREDENTIALS_FILE)`

`client = kms_v1.KeyManagementServiceClient(credentials = credentials)`

shermanyin commented on tools/gcp_kms_secrets.py:
```diff
@@ -0,0 +1,406 @@
+
+# Use argparse to determine user's desired action, encryption or decryption of terraform.tfvars
+def main():
+    parser = argparse.ArgumentParser(description="This script encrypts or decrypts the secrets  \
```
If we have a long string like this, I typically put it in a var and then just reference the var here instead of creating a multi-line function call. see the quickstart wrapper script.

shermanyin commented on tools/gcp_kms_secrets.py:
```diff
@@ -0,0 +1,406 @@
+                                    inside the user-specified terraform.tfstate file by         \
+                                    prompting the user for one of the three deployment types    \
+                                    (single, multi-region, dc-only). Select the mode by setting \
+                                    a flag: -e for encryption or -d for decryption")
```
I'm thinking it's probably more flexible to let the user specify which file they want to run the tool on instead of making them select 1, 2 or 3 and hard code the paths.  It's bad enough that we're hard-coding which values to encrypt now.  I'm debating whether we should move all the secrets to a separate tfvars file.  

So a user might run "tfvars-encrypt ../deployments/gcp/single-connector/terrafrom.tfvars" 

Suggestion: "Creates GCP KMS keyring and key, and use the key to encrypt secrets in terraform.tfvars specified. "

shermanyin commented on tools/gcp_kms_secrets.py:
```diff
@@ -0,0 +1,406 @@
+                                    (single, multi-region, dc-only). Select the mode by setting \
+                                    a flag: -e for encryption or -d for decryption")
+
+    parser.add_argument("-e", help="run this script in encryption mode", action='store_true')
```
There should be no flag needed for encrypt, which is the default. just -d for decrypt

shermanyin commented on tools/gcp_kms_secrets.py:
```diff
@@ -0,0 +1,406 @@
+    elif args.d:
+        print("Decryption mode selected...\n")
+        decrypt_tfvars_secrets()
+    else:
```
this would not be needed if the default is -e.

shermanyin commented on tools/gcp_kms_secrets.py:
```diff
@@ -0,0 +1,406 @@
+#!/usr/bin/python3
```
I know I don't do this either and I should, but read up on https://www.python.org/dev/peps/pep-0257/ and document your functions properly.

shermanyin replied:
A bit too many comments. I generally like comments as it explains clearly the author's intentions, but if it is very obvious then there doesn't need to be comments. e.g.
    # Create a key ring
    response = client.create_key_ring(parent, key_ring_id, keyring)
Also, try to document why you are doing something rather than what you are doing (unless it's not obvious). Like the above, you are obviously creating a key ring by calling "create_key_ring()"

shermanyin commented on tools/gcp_kms_secrets.py:
```diff
@@ -0,0 +1,406 @@
+            
+            # Split using the first delimiter
+            key, value = map(str.strip, line.split('=', 1))
+            tf_data[key] = value.replace("\"", "")
```
where do we run into this issue of having \"?

Edwin-Pau replied:
The above block of the code was my fix to the double quotes problem that I had a while back.

When I reused the quickstart_config_read function in your quickstart.py to read in lines, I made the mistake of using it without modifying it for this use (so it was grabbing the double quotes in tfvars secrets and treating it as part of the secrets when encrypting). 
`key, value = map(str.strip, line.split(':'))`
`cfg_data[key] = value`
I didn't think much of it the first time since I sanitized the double quotes during decryption output so I didn't think this was the issue. 

Lesson learned!

shermanyin commented on tools/gcp_kms_secrets.py:
```diff
@@ -0,0 +1,406 @@
+        "ad_service_account_password" : tfvars_dict.get("ad_service_account_password")
+    }
+    
+    if "single-connector" in tfvars_path or "multi-region" in tfvars_path:
```
don't think we should rely on the path.

how about just encrypt fields as we find them, and ignore if they are missing.

shermanyin commented on tools/gcp_kms_secrets.py:
```diff
@@ -0,0 +1,406 @@
+
+    if tfvars_dict.get("kms_keyring_name")   == None or \
+       tfvars_dict.get("kms_cryptokey_name") == None:
+        key_ring_id   = "terraform-keyring"
```
this kind of breaks the convention of adding the prefix to all resources created.

shermanyin commented on tools/gcp_kms_secrets.py:
```diff
@@ -0,0 +1,406 @@
+    project_id = tfvars_dict.get("gcp_project_id")
+    location   = "global"
+
+    if tfvars_dict.get("kms_keyring_name")   == None or \
```
I'm thinking maybe we don't even let users choose the name. just call them cas_keyring and cas_key or something.

Edwin-Pau replied:
Going with your suggestion of just creating the keyring and key for a better user experience, do I still want to add the user-specified prefix (if they used the field in terraform.tfvars) to this?

shermanyin commented on tools/gcp_kms_secrets.py:
```diff
@@ -0,0 +1,406 @@
+        crypto_key_id = tfvars_dict.get("kms_cryptokey_name")
+
+    # List all key rings
+    list_key_rings(client, project_id, location)
```
why do we need this?

shermanyin commented on tools/gcp_kms_secrets.py:
```diff
@@ -0,0 +1,406 @@
+        kms_keyring_id = create_key_ring(client, project_id, location, key_ring_id)
+        print("Created key ring: {}\n".format(kms_keyring_id))
+    
+    except Exception as err:
```
not sure what the logic here is. you're trying to take care of the case when the user-specified key ring and key already exist? If so I prefer if you explicitly check for it rather than try and catch the error, because there could be other errors other than "key already exist".

shermanyin commented on tools/gcp_kms_secrets.py:
```diff
@@ -0,0 +1,406 @@
+        for secret in secrets:
+            print("Encrypting {}...\n".format(secret))
+            ciphertext = encrypt_plaintext(client, project_id, location, key_ring_id, crypto_key_id, secrets.get(secret))
+            print("Finished encrypting {0}:\n{1}\n".format(secret, ciphertext))
```
don't need 0 and 1.  probably no need to print out the ciphertext either since we don't expect them to be legible or needs copying.

shermanyin commented on tools/gcp_kms_secrets.py:
```diff
@@ -0,0 +1,406 @@
+            
+    except Exception as err:
+        print("An exception occurred encrypting secrets:")
+        print("{}\n".format(err))
```
if an exception occur, shouldn't you exit?

shermanyin commented on tools/gcp_kms_secrets.py:
```diff
@@ -0,0 +1,406 @@
+
+# Write the new tfvars file using encrypted secrets
+def write_new_tfvars(tfvars_file, secrets, kms_cryptokey_id):
+    kms_resource_list  = kms_cryptokey_id.split('/')
```
best if you can give an example of why "5" and "7" are chosen, so reader doesn't have to google what the string is supposed to look like.

shermanyin commented on tools/gcp_kms_secrets.py:
```diff
@@ -0,0 +1,406 @@
+    
+    with open(tfvars_file, 'r') as f:
+        for line in f:
+            if "dc_admin_password" in line:
```
what if I have a comment line with "dc_admin_password" in it somewhere?

shermanyin commented on tools/gcp_kms_secrets.py:
```diff
@@ -0,0 +1,406 @@
+
+
+# Write the new tfvars file using encrypted secrets
+def write_new_tfvars(tfvars_file, secrets, kms_cryptokey_id):
```
need to save the original as a backup.  Log this file name to the output to remind the user where the backup is.

Edwin-Pau replied:
Do we still want the decryption function if we are saving original plaintext as backup?

shermanyin replied:
Good point :)  I guess it's not really needed if we have a backup.

Commit:
Refactor reading, encrypting, and writing tfvars

The functions that read, encrypt, and write the terraform.tfvars file
have been refactored by removing any hard-coded strings. The script
encrypts all secrets below the line "# Secrets below can be encrypted
using GCP-KMS, refer to the README for more information." and writes
to a new tfvars file. The original file is saved with a .backup postfix.

Signed-off-by: Edwin Pau <epau@teradici.com>

Commit:
Refactor to use class and attributes

Script has been refactored to use a class that represents a GCP KMS
client. The client has encryption and decrytion methods that can be
called to encrypt or decrypt a terraform.tfvars file. This design
pattern will allow easier support of AWS KMS in the future.

Signed-off-by: Edwin Pau <epau@teradici.com>

shermanyin commented on docs/README-gcp.md:
```diff
@@ -38,12 +38,22 @@ Login to Cloud Access Manager Admin Console at https://cam.teradici.com using a
-The Terraform scripts are designed to support both plaintext and KMS-encrypted secrets. Plaintext secrets requires no extra steps, but will be stored in plaintext in the above mentioned locations. It is recommended that secrets are first encrypted before being entered into terraform.tfvars.
+The Terraform scripts are designed to support both plaintext and KMS-encrypted secrets. Plaintext secrets requires no extra steps, but will be stored in plaintext in the above mentioned locations. It is recommended to encrypt the secrets in the terraform.tfvars file before deploying. Secrets can be encrypted manually first before being entered into terraform.tfvars, or they can be encrypted using ...
 
 To encrypt secrets using the KMS crypto key created above, follow the instructions here: https://cloud.google.com/kms/docs/encrypt-decrypt. Base64 encode the ciphertext before copying and pasting it into terraform.tfvars. For example, execute the following command in GCP Cloud Shell:
```
we can add a section header here called "Manual encryption"

shermanyin commented on docs/README-gcp.md:
````diff
@@ -38,12 +38,22 @@ Login to Cloud Access Manager Admin Console at https://cam.teradici.com using a
 
 ```echo -n <secret> | gcloud kms encrypt --location <location> --keyring <keyring_name> --key <key_name> --plaintext-file - --ciphertext-file - | base64```
 
+Alternatively, run the ```gcp_kms_secrets``` python3 script under the tools directory to automate the KMS encryption process. This script creates a key ring and crypto key on GCP KMS, then it encrypts all the secrets inside of your terraform.tfvars using this key. It will also encrypt any files that are entered as a path under the secrets line.
````
we can add a section header called "Encryption using Python script"

shermanyin commented on tools/kms_secrets_encryption.py:
```diff
@@ -0,0 +1,548 @@
+
+        # Create an API client for the KMS API using the provided GCP service account
+        self.credentials = service_account.Credentials.from_service_account_file(self.gcp_credentials_file)
+        self.client      = kms_v1.KeyManagementServiceClient(credentials = self.credentials)
```
i would call this kms_client

shermanyin commented on tools/kms_secrets_encryption.py:
```diff
@@ -0,0 +1,548 @@
+        self.gcp_credentials_file = self.tfvars_data.get("gcp_credentials_file")
+
+        # Create an API client for the KMS API using the provided GCP service account
+        self.credentials = service_account.Credentials.from_service_account_file(self.gcp_credentials_file)
```
i would call it gcp_credentials

shermanyin commented on tools/kms_secrets_encryption.py:
```diff
@@ -0,0 +1,548 @@
+
+        # GCP KMS resource variables
+        self.project_id      = self.tfvars_data.get("gcp_project_id")
+        self.location        = "global"
```
If it's just going to be a static "global", just make it a global variable at the top of the file and refer to that.

Edwin-Pau replied:
As per my explanation over Slack, I will be keeping this within the GCP instance since AWS in the future will not be using this location variable.

shermanyin commented on tools/kms_secrets_encryption.py:
```diff
@@ -0,0 +1,548 @@
+from google.oauth2 import service_account
+
+
+class GCP_KMS_Client:
```
gcp_kms_client sounds like what google.cloud.kms_v1 returns.  I'd call this something like tfvar_gcp_encryptor or something.

shermanyin commented on tools/kms_secrets_encryption.py:
```diff
@@ -0,0 +1,548 @@
+        parent   = self.client.location_path(self.project_id, self.location)
+        response = self.client.list_key_rings(parent)
+
+        # Extract just the key ring name from each response and put into a list
```
it's often helpful to give an examle of what the the output from list_key_rings() so that we know what you are trying to process.  e.g. "the response returns something like this: "some/path/keyname" and we're extracting the "keyname" only".

shermanyin commented on tools/kms_secrets_encryption.py:
```diff
@@ -0,0 +1,548 @@
+        key_rings_list = self.get_key_rings()
+
+        # Create the key ring only if it doesn't exist
+        if key_ring_id not in key_rings_list:
```
Good that you are checking if the key ring already exist.  I would print a message for the case when an existing key ring is used instead. "Using existing keyring {name}"

Also, a more pythonic way to do this is go ahead and try to create the key ring without checking, then presumably GCP will error and tell you the ring exist, which you should catch, print "using existing keyring {name}", then continue.  In that case, you can do all that in create_key_ring() so you don't need initialize_keyring().  You don't have to change this but just FYI

shermanyin commented on tools/kms_secrets_encryption.py:
```diff
@@ -0,0 +1,548 @@
+
+            # Extract just the crypto key name from each response and put that into a temp list
+            # [2] to get the string after the separator
+            temp_list = list(map(lambda key: key.name.rpartition('/')[2], response))
```
would be nice to include example of what "key" looks like before in the comments.

shermanyin commented on tools/kms_secrets_encryption.py:
```diff
@@ -0,0 +1,548 @@
+        """
+
+        key_rings_list   = self.get_key_rings()
+        crypto_keys_list = self.get_crypto_keys(key_rings_list)
```
why do we have to create a list of all keys from all key_rings?  aren't we only interested in keys within the same keyring?

shermanyin commented on tools/kms_secrets_encryption.py:
```diff
@@ -0,0 +1,548 @@
+        crypto_keys_list = self.get_crypto_keys(key_rings_list)
+
+        # Create a crypto key only if it doesn't exist
+        if crypto_key_id not in crypto_keys_list:
```
see comment on line 366. Same applies here

shermanyin commented on tools/kms_secrets_encryption.py:
```diff
@@ -0,0 +1,548 @@
+                if "# Secrets below can be encrypted using GCP-KMS, refer to the README for more information." in line:
+                    break
+
+                if line[0] in ('#', '\n'):
```
i would trim the line first, so that this also works if there are spaces before the first #

shermanyin commented on tools/kms_secrets_encryption.py:
```diff
@@ -0,0 +1,548 @@
+
+        with open(tfvars_file, 'r') as f:
+            for line in f:
+                if "# Secrets below can be encrypted using GCP-KMS, refer to the README for more information." in line:
```
hmm. can we have something shorter to indicate the separation? if someone accidentally added a space somewhere in the line then it's not going to work. Maybe something like # <--start of secrets section, do not edit this line. --> or something like that.

Also, whatever line you use, make it a global at the top of the file, since u'll use it more than once.

shermanyin commented on tools/kms_secrets_encryption.py:
```diff
@@ -0,0 +1,548 @@
+        
+            file_decrypted = self.decrypt_ciphertext(file_ciphertext)
+
+            file_path_decrypted = "{}".format(file_path).replace(".encrypted", "")
```
what if the filename doesn't have ".encrypted" in it?

shermanyin commented on tools/kms_secrets_encryption.py:
```diff
@@ -0,0 +1,548 @@
+        """
+
+        # UTF-8 encoding of plaintext
+        plaintext = plaintext.encode("utf-8")
```
this could be rolled into 274, no need to redefine plaintext

shermanyin commented on tools/kms_secrets_encryption.py:
```diff
@@ -0,0 +1,548 @@
+        response = self.client.encrypt(self.crypto_key_path, plaintext)
+
+        # Base64 encoding of ciphertext
+        ciphertext = base64.b64encode(response.ciphertext).decode()
```
do we need to do decode('utf-8')?

shermanyin commented on tools/kms_secrets_encryption.py:
```diff
@@ -0,0 +1,548 @@
+            print("Encrypting file: {}...".format(file_path))
+            
+            with open(file_path) as f:
+                f = json.load(f)
```
why do we need to do json.load() then json.dumps() after?

Edwin-Pau replied:
It didn't work if I only use dumps since it requires a JSON object, not a file object. I believe I need to use json.load() to convert a file object to json object, then use .dumps() to convert a json object to a string.
Error if I remove json.load():
`An exception occurred encrypting the file:
Object of type 'TextIOWrapper' is not JSON serializable`

shermanyin commented on tools/kms_secrets_encryption.py:
```diff
@@ -0,0 +1,548 @@
+        
+            f_string = json.dumps(f)
+            f_encrypted_string = self.encrypt_plaintext(f_string)
+            file_path_encrypted = "{}.encrypted".format(file_path)
```
Here is what I think you are doing. On encrypting a file, say "a.txt", you encrypt it and save it as "a.txt.encrypted", then rename "a.txt" to "a.txt.backup".  On decrypting, you expect the file to end in ".encrypted", say "b.txt.encrypted", you decrypt it and take out the .encrypted to get "b.txt", then you delete "b.txt.encrypted".
how about:
on encryption: take a.txt, encrypt it and call it a.txt.encrypted. leave a.txt alone
on decryption: take b.txt, decrypt it and call it b.txt.decrypted. leave b.txt alone

shermanyin commented on tools/kms_secrets_encryption.py:
```diff
@@ -0,0 +1,548 @@
+                    begin_reading_secrets = True
+                    continue
+                
+                if line[0] in ('#', '\n'):
```
trim spaces

shermanyin commented on tools/kms_secrets_encryption.py:
```diff
@@ -0,0 +1,548 @@
+                lines.append(line.rstrip())
+
+                # Append kms_cryptokey_id to tfvars if it is encrypting
+                if "# must be in the same project and have KMS decryptor permissions for this key." in line:
```
dun think it's a good idea to check for such a specific string. any updates to the comments will break this

shermanyin commented on tools/kms_secrets_encryption.py:
```diff
@@ -0,0 +1,548 @@
+                    break
+            
+            # Add all secrets back into tfvars
+            for key, value in self.tfvars_secrets.items():
```
wouldn't this change the order of the key value items? I'd try to keep the order the same. e.g. read each line and parse the key, if it needs updating then replace the value with the encrypted version. Reference how it was done in gcp-cloudshell-quickstart.py line 246

shermanyin commented on tools/kms_secrets_encryption.py:
```diff
@@ -0,0 +1,548 @@
+            for secret in self.tfvars_secrets:
+                # Additional handling needed if the string is a path to a file (IE. cam_credentials_file)
+                if os.path.isfile(self.tfvars_secrets.get(secret)):
+                    file_encrypted = self.encrypt_file(self.tfvars_secrets.get(secret))
```
dun really need the file_encrypted variable

shermanyin commented on tools/kms_secrets_encryption.py:
```diff
@@ -0,0 +1,548 @@
+                    self.tfvars_secrets[secret] = file_encrypted
+                else:
+                    print("Encrypting {}...".format(secret))
+                    ciphertext = self.encrypt_plaintext(self.tfvars_secrets.get(secret))
```
ciphertext not needed.

shermanyin commented on tools/kms_secrets_encryption.py:
```diff
@@ -0,0 +1,548 @@
+            raise SystemExit()
+
+        # Abort the decryption if GCP credentials is missing
+        if not self.tfvars_data.get("gcp_credentials_file"):
```
could move this check out to the parent since both encrypt and decrypt function calls this.

Commit:
Update terraform.tfvars samples for GCP

Sample files for terraform.tfvars were modified to ensure that secrets
can be detected by Tfvars_Encryptor_GCP after a specified line in the
file. All secrets below the line "# <-- Start of secrets section,
do not edit this line. -->" will be encrypted by the script. The default
for the kms_cryptokey_id line has also been changed to a comment.

Signed-off-by: Edwin Pau <epau@teradici.com>

shermanyin commented on docs/README-gcp.md:
````diff
@@ -38,12 +38,23 @@ Login to Cloud Access Manager Admin Console at https://cam.teradici.com using a
 ```echo -n <secret> | gcloud kms encrypt --location <location> --keyring <keyring_name> --key <key_name> --plaintext-file - --ciphertext-file - | base64```
 
+#### Encryption Using Python Script
+Alternatively, run the kms_secrets_encryption.py Python 3 script under the tools directory to automate the KMS encryption process. 
````
need a step here.  "first, enter the plaintext secrets into terraform.tfvars, save the file, then run the script"

shermanyin commented on docs/README-gcp.md:
````diff
@@ -38,12 +38,23 @@ Login to Cloud Access Manager Admin Console at https://cam.teradici.com using a
+
+The script rewrites your existing terraform.tfvars secrets using ciphertext so that it is ready to be used for your Terraform deployment. In addition, any text files specified under the secrets as a path will also be encrypted so that it is ready to be used by Terraform. The script will also update the path to the encrypted text file inside of the new terraform.tfvars.
+
+The default mode is encryption; to decrypt the secrets back to plaintext, use the ```-d``` flag when executing the script. Decryption mode decrypts the ciphertext in the terraform.tfvars file back to the original plaintext. The script rewrites the ciphertext version of terraform.tfvars file with the plaintext secrets again. 
````
A bit wordy.

The script can also reverse the encryption by running with the '-d' flag.  See script's documentation for details (--help)

shermanyin commented on tools/README.md:
````diff
@@ -11,3 +11,25 @@ pip install names
+
+To decrypt:
+```
+./kms_secrets_encryption.py <path/to/terraform.tfvars> -d
````
usually the -d comes before the filename.

shermanyin commented on tools/kms_secrets_encryption.py:
```diff
@@ -0,0 +1,551 @@
+        return key_rings_list
+
+
+    def read_tfvars_data(self, tfvars_file):
```
This and the secret reading function could be 1 function since they share so much in common.
